secant/Features/BackupFlow/RecoveryPhraseValidation.swift

```diff
@@ -0,0 +1,231 @@
+                var newMissingWords = missingWordChips
+                newMissingWords[missingChipIndex] = .empty
+
+                var newCompletionState = Array(validationWords)
+                newCompletionState.append(ValidationWord(groupIndex: group, word: word))
+
+                return RecoveryPhraseValidationState(
+                    phrase: phrase,
+                    missingIndices: missingIndices,
+                    missingWordChips: newMissingWords,
+                    validationWords: newCompletionState
+                )
+            } else {
+                var newCompletion = validationWords
+                newCompletion.append(ValidationWord(groupIndex: group, word: word))
+
+                return RecoveryPhraseValidationState(
+                    phrase: phrase,
+                    missingIndices: missingIndices,
+                    missingWordChips: Array(repeating: .empty, count: RecoveryPhraseValidationState.phraseChunks),
+                    validationWords: newCompletion
+                )
+            }
+        case .complete:
+            return self
+        }
+    }
+
+    static func randomIndices() -> [Int] {
+        return (0..<phraseChunks).map { _ in
+            Int.random(in: 0 ..< wordGroupSize)
+        }
+    }
+}
+
+extension RecoveryPhrase.Chunk {
+    /// Returns an array of words where the word at the missing index will be an empty string
+    func words(with missingIndex: Int) -> [String] {
+        assert(missingIndex >= 0)
+        assert(missingIndex < self.words.count)
+        var wordsApplyingMissing = self.words
+        wordsApplyingMissing[missingIndex] = ""
+        return wordsApplyingMissing
+    }
+}
+
+enum RecoveryPhraseValidationAction: Equatable {
+    case updateRoute(RecoveryPhraseValidationState.Route?)
+    case reset
+    case drag(wordChip: PhraseChip.Kind, intoGroup: Int)
+    case succeed
+    case fail
+    case proceedToHome
+    case displayBackedUpPhrase
+}
+
+typealias RecoveryPhraseValidationReducer = Reducer<RecoveryPhraseValidationState, RecoveryPhraseValidationAction, BackupPhraseEnvironment>
+
+extension RecoveryPhraseValidationReducer {
+    static let `default` = RecoveryPhraseValidationReducer { state, action, environment in
+        switch action {
+        case .reset:
+            state = RecoveryPhraseValidationState.random(phrase: state.phrase)
+            return .none
+
+        case let .drag(wordChip, group):
+            state = state.apply(chip: wordChip, into: group)
+
+            // Trigger a delayed effect to proceed to the next step
+            if case .complete = state.step {
+                if state.isValid {
+                    return Effect(value: .succeed).delay(for: 1, scheduler: environment.mainQueue).eraseToEffect()
+                } else {
+                    return Effect(value: .fail).delay(for: 3, scheduler: environment.mainQueue).eraseToEffect()
+                }
+            }
+            return .none
```

Comment:
I was poking around in code with this reducer helper function, and I was able to clean up the function a bit and get rid of the intermediate state, which reduces the overall code by a fair amount.  If we just put this state manipulation code right in the reducer where we have access to inout state, all the state manipulations are now done directly in the reducer, and we don't need functions that produce new state where we have to overwrite the entire state value.

Your reducer grows by 19 lines of code but you get to get rid of a 50+ line of code function that contains state manipulation logic that is probably better off right here in the reducer, that is the reducers job after all.

There are a few reasons this seems very clean to me.  
1) state manipulations stay in the reducer itself
2) the Values we are working with are being widdled down to just representations of a Value in time, without mutating funcs or funcs producing new states.
3) In the future, if we want to read through any state manipulations in our store, they all happen right here in the reducer.

```suggestion
            guard case let .unassigned(word) = wordChip else { return .none }

            switch state.step {
            case .initial:
                guard let missingChipIndex = state.missingWordChips.firstIndex(of: wordChip) else { break }

                state.missingWordChips[missingChipIndex] = .empty
                state.validationWords = [ValidationWord(groupIndex: group, word: word)]

            case .incomplete:
                guard let missingChipIndex = state.missingWordChips.firstIndex(of: wordChip) else { break }

                if state.validationWords.count < (RecoveryPhraseValidationState.phraseChunks - 1) {
                    state.missingWordChips[missingChipIndex] = .empty
                    state.validationWords.append(ValidationWord(groupIndex: group, word: word))
                } else {
                    state.missingWordChips = Array(
                        repeating: .empty,
                        count: RecoveryPhraseValidationState.phraseChunks
                    )
                    state.validationWords.append(ValidationWord(groupIndex: group, word: word))
                }

            case .complete:
                if state.isValid {
                    return Effect(value: .succeed)
                        .delay(for: 1, scheduler: environment.mainQueue)
                        .eraseToEffect()
                } else {
                    return Effect(value: .fail)
                        .delay(for: 3, scheduler: environment.mainQueue)
                        .eraseToEffect()
                }
            }
            return .none
```

Reply:
Thanks for pointing this out. I'm implementing this change.

Reply:
pushed